File: framework/backend/go/k1s0-domain-event/envelope.go
package domainevent

import (
	"encoding/json"
	"time"

	"github.com/google/uuid"
)

// EventMetadata holds metadata associated with a domain event.
type EventMetadata struct {
	// EventID is a unique identifier for this event.
	EventID string `json:"event_id"`

	// OccurredAt is the time the event occurred.
	OccurredAt time.Time `json:"occurred_at"`

	// Source is the name of the service that produced the event.
	Source string `json:"source"`

	// CorrelationID is used for distributed tracing (optional).
	CorrelationID string `json:"correlation_id,omitempty"`

	// CausationID tracks causality (optional).
	CausationID string `json:"causation_id,omitempty"`
}

// NewEventMetadata creates a new EventMetadata with a generated ID and current time.
func NewEventMetadata(source string) EventMetadata {
	return EventMetadata{
		EventID:    uuid.New().String(),
		OccurredAt: time.Now().UTC(),
		Source:     source,
	}
}

// EventEnvelope wraps an event payload with metadata.
type EventEnvelope struct {
	// EventType is the event type identifier.
	EventType string `json:"event_type"`

	// Metadata contains event metadata.
	Metadata EventMetadata `json:"metadata"`

	// Payload is the JSON-encoded event body.
	Payload json.RawMessage `json:"payload"`
}

// NewEventEnvelope creates an EventEnvelope from a DomainEvent.
func NewEventEnvelope(event DomainEvent, source string) (*EventEnvelope, error) {
	payload, err := json.Marshal(event)
	if err != nil {
		return nil, err
	}
	return &EventEnvelope{
		EventType: event.EventType(),
		Metadata:  NewEventMetadata(source),
		Payload:   payload,
	}, nil
}

File: framework/backend/python/packages/k1s0-domain-event/src/k1s0_domain_event/envelope.py
"""Event envelope and metadata for wrapping domain events."""

from __future__ import annotations

import json
from datetime import UTC, datetime
from uuid import UUID, uuid4

from pydantic import BaseModel, Field

from k1s0_domain_event.event import DomainEvent


class EventMetadata(BaseModel):
    """Metadata attached to every domain event envelope."""

    event_id: UUID = Field(default_factory=uuid4)
    occurred_at: datetime = Field(default_factory=lambda: datetime.now(UTC))
    source: str
    correlation_id: str | None = None
    causation_id: str | None = None


class EventEnvelope(BaseModel):
    """Wraps a domain event with metadata and a serialized payload."""

    event_type: str
    metadata: EventMetadata
    payload: str

    @classmethod
    def wrap(
        cls,
        event: DomainEvent,
        source: str,
        *,
        correlation_id: str | None = None,
        causation_id: str | None = None,
    ) -> EventEnvelope:
        """Create an envelope from a domain event.

        Args:
            event: The domain event to wrap.
            source: The source service or component name.
            correlation_id: Optional correlation ID for tracing.
            causation_id: Optional causation ID for tracing.

        Returns:
            A new EventEnvelope containing the serialized event.
        """
        payload = json.dumps(
            {
                "aggregate_id": event.aggregate_id(),
                "aggregate_type": event.aggregate_type(),
            }
        )
        metadata = EventMetadata(
            source=source,
            correlation_id=correlation_id,
            causation_id=causation_id,
        )
        return cls(
            event_type=event.event_type(),
            metadata=metadata,
            payload=payload,
        )
